### PR TITLE
Synchronize MultiObjectSMREntry

### PR DESCRIPTION
**Pull request by Maithem**
Synchronize reads and writes on a MultiObjectSMREntry.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **dfa6110**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #854 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/PxalEQIOGlLQkQ83i02XkYYyBy3XC5Pf)

**Comment by slfritchie** [comment]
This looks like it's trying to address #853.

**Comment by Maithem** [comment]
@slfritchie Yes, I think its a timing issue thats related to memory, I don't think I can write an explicit  test case. I believe it happens by the 3rd get. Also, apparently current is not called unless that list is not empty ?

**Comment by Maithem** [comment]
@slfritchie Also, I dont always see a ConcurrentModificationException, Its mostly seems that stale reads of that map are causing the out of bound exception

**Comment by slfritchie** [comment]
Hm, are there two problems here?

1. Stale read. This PR seems good but I can't figure out if its sufficient for all stale read cases.
2. `getWriteSetEntryList()` returning an empty list, which is what I've interpreted #853 to be.  If returning an empty list is possible, then this PR isn't enough to avoid WriteSetSMRStream::current throwing IOOB exceptions.

**Comment by Maithem** [comment]
@slfritchie It seems like other places of the code expect it to return a null, if the index doesnt exist, but also returning a null would mean throwing a runtime exception NoRollbackException(stream.pos(), rollbackVersion);

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java [hunk 1]
```diff
@@ -167,10 +167,17 @@ void mergeTransaction() {
         if (Address.nonAddress(currentContextPos)) {
             currentContextPos = -1;
         }
-        return Collections.singletonList(contexts
-                .get(currentContext)
-                .getWriteSetEntryList(id)
-                .get((int)(currentContextPos)));
+
+        try {
+            return Collections.singletonList(contexts
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java]: remove

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **a61f79d**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #854 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/a8MtAJek1zra7WpVrh3FCilxBa2VZcsU)

**Comment by slfritchie** [comment]
Did your work last night succeed in finding the reason for why synchronizing the map in this class works as it does?

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/MultiObjectSMREntry.java [hunk 2]
```diff
@@ -28,11 +28,15 @@
 public class MultiObjectSMREntry extends LogEntry implements ISMRConsumable {
 
     // map from stream-ID to a list of updates encapsulated as MultiSMREntry
-    @Getter
-    public Map<UUID, MultiSMREntry> entryMap = new HashMap<>();
+    private Map<UUID, MultiSMREntry> entryMap;
+
+    synchronized public Map<UUID, MultiSMREntry> getEntryMap(){
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/MultiObjectSMREntry.java]: do we really need the getter for this map anymore ?

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java [hunk 3]
```diff
@@ -167,10 +167,17 @@ void mergeTransaction() {
...
+                    .get(currentContext)
+                    .getWriteSetEntryList(id)
+                    .get((int) (currentContextPos)));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java]: remove these .

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java [hunk 4]
```diff
@@ -167,10 +167,17 @@ void mergeTransaction() {
...
         }
-        return Collections.singletonList(contexts
-                .get(currentContext)
-                .getWriteSetEntryList(id)
-                .get((int)(currentContextPos)));
+
+        try {
+            return Collections.singletonList(contexts
+                    .get(currentContext)
+                    .getWriteSetEntryList(id)
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/WriteSetSMRStream.java]: if I were to tighten this up, I would just have getWriteSetEntry(id, currentContextPost) or something to that effect. This leaves a lose end for other bugs.

**Comment by Maithem** [comment]
@slfritchie Yes, it was what I described earlier. You have a reader and writer in two different threads and access is no synchronized, so some of the reads were causing an out of bound exception

**Comment by Maithem** [comment]
Addressed all comments.

**Pull request merged**